Potential permission errors occur in ckpt_save when multiple user are using the same socket file.

When **two different users** are using `dlrover-run` to train their models. Ckpt saver creates sock files with the same name in the same place.  However, this can cause permission errors as follows: 

```log
[2023-12-17 10:23:54,547] [INFO] [trainer.py:435:_set_gradient_accumulation_steps] Rank = 0, World size = 1, Gradient accumulation steps = 1
Traceback (most recent call last):
  File "/home/lan/CodeSpace/dlrover_space/dlrover/examples/pytorch/nanogpt/train.py", line 525, in <module>
    train()
  File "/home/lan/CodeSpace/dlrover_space/dlrover/examples/pytorch/nanogpt/train.py", line 310, in train
    ckpt_manager = CheckpointManger.init_checkpoint_manager(
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/site-packages/dlrover/trainer/torch/elastic/checkpoint.py", line 201, in init_checkpoint_manager
    return DDPCheckpointManger(
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/site-packages/dlrover/trainer/torch/elastic/checkpoint.py", line 303, in __init__
    super().__init__(
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/site-packages/dlrover/trainer/torch/elastic/checkpoint.py", line 245, in __init__
    self._ckpt_engine = NoShardingCheckpointEngine(
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/site-packages/dlrover/python/elastic_agent/torch/ckpt_saver.py", line 1115, in __init__
    super().__init__(checkpoint_dir)
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/site-packages/dlrover/python/elastic_agent/torch/ckpt_saver.py", line 950, in __init__
    self._update_saver_config()
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/site-packages/dlrover/python/elastic_agent/torch/ckpt_saver.py", line 995, in _update_saver_config
Exception in thread checkpoint-saver:
Traceback (most recent call last):
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/site-packages/dlrover/python/elastic_agent/torch/ckpt_saver.py", line 398, in _save
    saver: CheckpointSaver = class_def(**class_meta.init_args)
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/site-packages/dlrover/python/elastic_agent/torch/ckpt_saver.py", line 372, in __init__
    self._event_queue = SharedQueue(name=qname, create=True)
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/site-packages/dlrover/python/common/multi_process.py", line 341, in __init__
    super().__init__(name, create)
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/site-packages/dlrover/python/common/multi_process.py", line 167, in __init__
    self._init_socket()
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/site-packages/dlrover/python/common/multi_process.py", line 183, in _init_socket
    self._server = _create_socket_server(self._socket_file)
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/site-packages/dlrover/python/common/multi_process.py", line 47, in _create_socket_server
    os.unlink(path)
PermissionError: [Errno 1] Operation not permitted: '/tmp/sharedqueue_checkpoint_lock_rank_0.sock'
    self._event_queue.put(event)
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/site-packages/dlrover/python/common/multi_process.py", line 382, in put
    self._request(request)
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/site-packages/dlrover/python/common/multi_process.py", line 197, in _request
    Exception ignored in: <function CheckpointSaver.__del__ at 0x7faddc3e43a0>
Traceback (most recent call last):
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/site-packages/dlrover/python/elastic_agent/torch/ckpt_saver.py", line 382, in __del__
    self.close()
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/site-packages/dlrover/python/elastic_agent/torch/ckpt_saver.py", line 438, in close
    if self._shm_handlers[i]:
IndexError: list index out of range
client = _create_socket_client(self._socket_file)
  File "/home/lan/DevelopmentTools/miniconda3/envs/dl_torch/lib/python3.8/site-packages/dlrover/python/common/multi_process.py", line 65, in _create_socket_client
    client.connect(path)
PermissionError: [Errno 13] Permission denied
ERROR:torch.distributed.elastic.multiprocessing.api:failed (exitcode: 1) local_rank: 0 (pid: 91778) of binary: /home/lan/DevelopmentTools/miniconda3/envs/dl_torch/bin/python
[2023-12-17 10:23:55,595] [ERROR] [training.py:548:_invoke_run] The worker fails with {0: ProcessFailure(local_rank=0, pid=91778, exitcode=1, error_file='<N/A>', error_file_data={'message': '<NONE>'}, message='To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html', timestamp=1702808635)}
[2023-12-17 10:23:55,596] [ERROR] [error_monitor.py:54:_handle_process_error] worker-0 on None restart 0 fails: {"0": {"local_rank": 0, "exitcode": 1, "message": "To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html", "datetime": "2023-12-17 10:23:55"}}
```

The sock files look like `srwxr-xr-x 1 root root 0 Dec 17 10:07 /tmp/sharedqueue_checkpoint_lock_rank_0.sock` after **ckpt_saver** from root user created the sock files. **They are not accessible to the other users.**

--- a/dlrover/python/common/constants.py
+++ b/dlrover/python/common/constants.py
@@ -209,6 +209,9 @@ class NodeEnv(object):
     WORKER_NUM = "WORKER_NUM"
     WORKER_RANK = "WORKER_RANK"
 
+    # process env
+    TORCHELASTIC_RUN_ID = "TORCHELASTIC_RUN_ID"
+
 
 class DatasetType(object):
     TEXT = "text"

--- a/dlrover/python/common/multi_process.py
+++ b/dlrover/python/common/multi_process.py
@@ -173,6 +173,12 @@ class LocalSocketComm(metaclass=ABCMeta):
     def _create_socket_path(self):
         """Create a file path for the local socket."""
         fname = self.__class__.__name__.lower() + "_" + self._name + ".sock"
+        job_name = os.getenv("JOB_NAME", "")
+        if job_name:
+            root_dir = os.path.join(TMP_DIR, job_name)
+        else:
+            root_dir = TMP_DIR
+        os.makedirs(root_dir, exist_ok=True)
         return os.path.join(TMP_DIR, fname)
 
     def _init_socket(self):

--- a/dlrover/python/common/multi_process.py
+++ b/dlrover/python/common/multi_process.py
@@ -25,6 +25,7 @@ from typing import Dict
 
 import _posixshmem
 
+from .constants import NodeEnv
 from .log import default_logger as logger
 
 TMP_DIR = "/tmp"
@@ -173,7 +174,7 @@ class LocalSocketComm(metaclass=ABCMeta):
     def _create_socket_path(self):
         """Create a file path for the local socket."""
         fname = self.__class__.__name__.lower() + "_" + self._name + ".sock"
-        job_name = os.getenv("JOB_NAME", "")
+        job_name = os.getenv(NodeEnv.TORCHELASTIC_RUN_ID, "")
         if job_name:
             root_dir = os.path.join(TMP_DIR, job_name)
         else:

--- a/dlrover/python/common/multi_process.py
+++ b/dlrover/python/common/multi_process.py
@@ -179,7 +179,7 @@ class LocalSocketComm(metaclass=ABCMeta):
         else:
             root_dir = TMP_DIR
         os.makedirs(root_dir, exist_ok=True)
-        return os.path.join(TMP_DIR, fname)
+        return os.path.join(root_dir, fname)
 
     def _init_socket(self):
         """Initialze a socket server."""

--- a/dlrover/python/elastic_agent/torch/ckpt_saver.py
+++ b/dlrover/python/elastic_agent/torch/ckpt_saver.py
@@ -27,7 +27,7 @@ import torch
 import torch.distributed as dist
 
 from dlrover.python.common import env_utils
-from dlrover.python.common.constants import CheckpointConstant
+from dlrover.python.common.constants import CheckpointConstant, NodeEnv
 from dlrover.python.common.log import default_logger as logger
 from dlrover.python.common.multi_process import (
     SharedDict,
@@ -260,7 +260,13 @@ class SharedMemoryHandler(object):
         self._buffer_size = 0
         meta_name = _CKPT_META_NAME_PREFIX + str(local_rank)
         self._tensor_meta = SharedDict(name=meta_name, create=host)
-        self._shm_name = _TENSOR_SHM_NAME_PREFIX + str(local_rank)
+        job_name = os.getenv(NodeEnv.JOB_NAME, "")
+        if job_name:
+            self._shm_name = (
+                job_name + "_" + _TENSOR_SHM_NAME_PREFIX + str(local_rank)
+            )
+        else:
+            self._shm_name = _TENSOR_SHM_NAME_PREFIX + str(local_rank)
         self._tensor_shm = None
 
     def close(self):

--- a/dlrover/python/elastic_agent/torch/ckpt_saver.py
+++ b/dlrover/python/elastic_agent/torch/ckpt_saver.py
@@ -260,7 +260,7 @@ class SharedMemoryHandler(object):
         self._buffer_size = 0
         meta_name = _CKPT_META_NAME_PREFIX + str(local_rank)
         self._tensor_meta = SharedDict(name=meta_name, create=host)
-        job_name = os.getenv(NodeEnv.JOB_NAME, "")
+        job_name = os.getenv(NodeEnv.TORCHELASTIC_RUN_ID, "")
         if job_name:
             self._shm_name = (
                 job_name + "_" + _TENSOR_SHM_NAME_PREFIX + str(local_rank)

--- a/dlrover/python/tests/test_ckpt_saver.py
+++ b/dlrover/python/tests/test_ckpt_saver.py
@@ -84,7 +84,7 @@ class ShardingEngineDemo(NoShardingCheckpointEngine):
 class SharedMemoryHandlerTest(unittest.TestCase):
     def setUp(self):
         local_rank = 1
-        os.environ["JOB_NAME"] = "unittest"
+        os.environ[NodeEnv.TORCHELASTIC_RUN_ID] = "unittest"
         SharedDict(_CKPT_META_NAME_PREFIX + str(local_rank), create=True)
         self._shm_handler = SharedMemoryHandler(local_rank, host=False)
 

--- a/dlrover/python/tests/test_ckpt_saver.py
+++ b/dlrover/python/tests/test_ckpt_saver.py
@@ -84,6 +84,7 @@ class ShardingEngineDemo(NoShardingCheckpointEngine):
 class SharedMemoryHandlerTest(unittest.TestCase):
     def setUp(self):
         local_rank = 1
+        os.environ["JOB_NAME"] = "unittest"
         SharedDict(_CKPT_META_NAME_PREFIX + str(local_rank), create=True)
         self._shm_handler = SharedMemoryHandler(local_rank, host=False)
 

--- a/dlrover/python/tests/test_multi_process.py
+++ b/dlrover/python/tests/test_multi_process.py
@@ -25,7 +25,7 @@ from dlrover.python.common.multi_process import (
 class SharedLockTest(unittest.TestCase):
     def test_shared_lock(self):
         name = "test"
-        os.environ["JOB_NAME"] = "test_job"
+        os.environ["TORCHELASTIC_RUN_ID"] = "test_job"
         server_lock = SharedLock(name, create=True)
         self.assertTrue(os.path.exists("/tmp/test_job/sharedlock_test.sock"))
         client_lock = SharedLock(name, create=False)

--- a/dlrover/python/tests/test_multi_process.py
+++ b/dlrover/python/tests/test_multi_process.py
@@ -11,6 +11,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import os
 import unittest
 
 from dlrover.python.common.multi_process import (
@@ -24,7 +25,9 @@ from dlrover.python.common.multi_process import (
 class SharedLockTest(unittest.TestCase):
     def test_shared_lock(self):
         name = "test"
+        os.environ["JOB_NAME"] = "test_job"
         server_lock = SharedLock(name, create=True)
+        self.assertTrue(os.path.exists("/tmp/test_job/sharedlock_test.sock"))
         client_lock = SharedLock(name, create=False)
         acquired = server_lock.acquire()
         self.assertTrue(acquired)

--- a/dlrover/trainer/tests/torch/elastic_run_test.py
+++ b/dlrover/trainer/tests/torch/elastic_run_test.py
@@ -23,7 +23,7 @@ from dlrover.trainer.torch.elastic_run import (
 
 class ElasticRunTest(unittest.TestCase):
     def test_launch_local_master(self):
-        handler, addr = _launch_dlrover_local_master("")
+        handler, addr = _launch_dlrover_local_master("", "test")
         available = _check_dlrover_master_available(addr)
         self.assertTrue(available)
         handler.close()

--- a/dlrover/trainer/torch/elastic_run.py
+++ b/dlrover/trainer/torch/elastic_run.py
@@ -179,7 +179,7 @@ class elastic_launch:
             )
 
 
-def _launch_dlrover_local_master(master_addr):
+def _launch_dlrover_local_master(master_addr, job_name):
     """Launch a subprocess to run the DLrover master."""
     logger.info(f"Start dlrover master with addr {master_addr}")
     if not master_addr:
@@ -189,8 +189,6 @@ def _launch_dlrover_local_master(master_addr):
         host = master_addr.split(":")[0]
         port = int(master_addr.split(":")[1])
     cmd = os.getenv("PYTHON_EXEC", sys.executable)
-    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S_%f")
-    job_name = f"standalone_{timestamp}"
     args = (
         "-u",
         "-m",
@@ -202,8 +200,6 @@ def _launch_dlrover_local_master(master_addr):
         "--platform",
         "local",
     )
-    if NodeEnv.JOB_NAME not in os.environ:
-        os.environ[NodeEnv.JOB_NAME] = job_name
     handler = SubprocessHandler(cmd, args, {}, "", "")
     dlrover_master_addr = f"{host}:{port}"
     return handler, dlrover_master_addr
@@ -245,9 +241,14 @@ def run(args):
     master_addr = os.getenv(NodeEnv.DLROVER_MASTER_ADDR, "")
     use_dlrover_launch = False
     node_rank = env_utils.get_node_rank()
+    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S_%f")
+    job_name = os.getenv(NodeEnv.JOB_NAME, f"standalone_{timestamp}")
+    os.environ[NodeEnv.TORCHELASTIC_RUN_ID] = job_name
     if args.standalone and node_rank == 0:
         # Only start the dlrover master on the rank-0 node.
-        master_handler, master_addr = _launch_dlrover_local_master(master_addr)
+        master_handler, master_addr = _launch_dlrover_local_master(
+            master_addr, job_name
+        )
         os.environ[NodeEnv.DLROVER_MASTER_ADDR] = master_addr
     if _check_dlrover_master_available(master_addr):
         use_dlrover_launch = True
@@ -268,6 +269,7 @@ def run(args):
         )
 
     config, cmd, cmd_args = _elastic_config_from_args(args)
+    config.run_id = job_name
     try:
         elastic_launch(
             config=config,

--- a/dlrover/trainer/torch/elastic_run.py
+++ b/dlrover/trainer/torch/elastic_run.py
@@ -78,6 +78,7 @@ import sys
 import telnetlib
 import time
 import uuid
+from datetime import datetime
 from typing import Callable, List, Tuple, Union
 
 from torch.distributed.argparse_util import check_env, env
@@ -188,6 +189,8 @@ def _launch_dlrover_local_master(master_addr):
         host = master_addr.split(":")[0]
         port = int(master_addr.split(":")[1])
     cmd = os.getenv("PYTHON_EXEC", sys.executable)
+    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S_%f")
+    job_name = f"standalone_{timestamp}"
     args = (
         "-u",
         "-m",
@@ -195,10 +198,12 @@ def _launch_dlrover_local_master(master_addr):
         "--port",
         f"{port}",
         "--job_name",
-        "dlrover-training",
+        job_name,
         "--platform",
         "local",
     )
+    if NodeEnv.JOB_NAME not in os.environ:
+        os.environ[NodeEnv.JOB_NAME] = job_name
     handler = SubprocessHandler(cmd, args, {}, "", "")
     dlrover_master_addr = f"{host}:{port}"
     return handler, dlrover_master_addr

